pull_request.py

```diff
@@ -50,35 +58,96 @@ def days_since_pushed(self) -> int:
   def days_since_changed(self) -> int:
     return util.days_since(self.last_changed_ts())
 
+  def _calculate_diff_at_commit(self, commit: str) -> str:
+    # Determine what changes this commit makes relative to master.
+    #
+    # I don't know a git command directly for it, so instead make a temporary
+    # repo, load all relevant commits into the repo, merge the commit in
+    # question into master, and then diff against origin/master.
+
+    print('Calculating diff at %s' % commit)
+    original_working_directory = os.getcwd()
+    tmp_repo_fname = 'tmp-repo'
+    try:
+      # Make a new clone to work in.
+      subprocess.check_call([
+          'git', 'clone',
+          'https://github.com/%s.git' % self._repo, tmp_repo_fname])
+      os.chdir(tmp_repo_fname)
+      remote_name = self.author()
+
+      # We can't refer to commit until we download it.
+      subprocess.check_call(['git', 'remote', 'add', remote_name, self._clone_url])
+      subprocess.check_call(['git', 'fetch', remote_name, self._ref])
+      subprocess.check_call(['git', 'merge', '--no-edit', commit])
+
+      completed_process = subprocess.run(
+        ['git', 'diff', 'origin/master'], stdout=subprocess.PIPE)
+      if completed_process.returncode != 0:
+        raise Exception(completed_process)
+
+      return completed_process.stdout.decode('utf-8')
+
+    except Exception:
+      print('Failed to get diff at %s' % commit)
+      return ''
+
+    finally:
+      # This runs even if we return above, and puts us back in the state we
+      # were in before we tried to calculate the diff.
+
+      os.chdir(original_working_directory)
+      if os.path.exists(tmp_repo_fname):
+        shutil.rmtree(tmp_repo_fname)
+
+  def _diff_at_commit(self, commit: str) -> str:
+    if commit not in self._commit_diffs:
```

Comment:
This is cool, but maybe we should reverse the order in which we process the reviews (from newest to oldest), so that we don't run extra/redundant diff checks? I think it would also make the flow easier to follow in `_calculate_reviews`.

Reply:
Good idea; that makes much more sense.

It does mean needing to store json a bit longer, but that isn't a problem at all.

Reply:
done in 0e10875